main/inc/lib/hook/HookEvent.class.php

```diff
@@ -0,0 +1,209 @@
+<?php
+/* For licensing terms, see /license.txt */
+
+
+abstract class HookEvent implements HookEventInterface
+{
+    public $observers;
+    public $eventName;
+    public $eventData;
+
+    /**
+     * Construct Method
+     * @param $eventName
+     * @throws Exception
+     */
+    protected function __construct($eventName)
+    {
+        if (self::isHookPluginActive()) {
+            $this->observers = new SplObjectStorage();
+            $this->eventName = $eventName;
+            $this->eventData = array();
+            $this->plugin = HookManagement::create();
+            $this->loadAttachments();
+        } else {
+            throw new \Exception('Hook Management Plugin is not active');
+        }
+    }
+
+    /**
+     * Return the singleton instance of Hook event.
+     * If Hook Management plugin is not enabled, will return NULL
+     * @return HookEventInterface|null
+     */
+    public static function create()
+    {
+        static $result = null;
+
+        if ($result) {
+            return $result;
+        } else {
+            try {
+                $class = get_called_class();
+                return new $class;
+            } catch (Exception $e) {
+                return null;
+            }
+        }
+    }
+
+    /**
+     * Attach an HookObserver
+     * @link http://php.net/manual/en/splsubject.attach.php
+     * @param \HookObserverInterface| $observer <p>
+     * The <b>HookObserver</b> to attach.
+     * </p>
+     * @return void
+     */
+    public function attach(HookObserverInterface $observer)
+    {
+        global $_hook;
+        $observerClass = get_class($observer);
+        $_hook[$this->eventName][$observerClass] = array(
+            'class_name' => $observerClass,
+            'path' => $observer->getPath(),
+            'plugin_name' => $observer->getPluginName(),
+        );
+        $this->observers->attach($observer);
+        $this->plugin->insertHook($this->eventName, $observerClass, HOOK_TYPE_ALL);
+    }
+
+    /**
+     * Detach an HookObserver
+     * @link http://php.net/manual/en/splsubject.detach.php
+     * @param \HookObserverInterface| $observer <p>
+     * The <b>HookObserver</b> to detach.
+     * </p>
+     * @return void
+     */
+    public function detach(HookObserverInterface $observer)
+    {
+        global $_hook;
+        $observerClass = get_class($observer);
+        unset($_hook[$this->eventName][$observerClass]);
+        $this->observers->detach($observer);
+        $this->plugin->deleteHook($this->eventName, $observerClass, HOOK_TYPE_ALL);
+    }
+
+    /**
+     * (PHP 5 &gt;= 5.1.0)<br/>
```

Comment:
That's OK. Chamilo requires a *minimum* of PHP 5.3, so you can remove the comment